remove-data/src/main/java/eu/europeana/metis/remove/discover/ExecutionPluginForest.java

```diff
@@ -0,0 +1,196 @@
+package eu.europeana.metis.remove.discover;
+
+import eu.europeana.metis.core.workflow.WorkflowExecution;
+import eu.europeana.metis.core.workflow.plugins.AbstractMetisPlugin;
+import eu.europeana.metis.core.workflow.plugins.PluginStatus;
+import eu.europeana.metis.core.workflow.plugins.PluginType;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+class ExecutionPluginForest {
+
+  private final Map<String, ExecutionPluginNode> nodesById = new HashMap<>();
+  private final Map<PluginLinkByTypeAndTimestamp, ExecutionPluginNode> nodesByTypeAndTimestamp;
+  private final List<ExecutionPluginNode> rootNodes;
+
+  /**
+   * These are non-link checking nodes with only link checking descendants.
+   */
+  private final Set<String> leafNodeIds;
+
+  ExecutionPluginForest(List<WorkflowExecution> executions) {
+    this(executions, Collections.emptySet());
+  }
+
+  ExecutionPluginForest(List<WorkflowExecution> executions, Set<String> ignoredPluginIds) {
+
+    // Convert all executions and plugins to nodes and add them to the node map
+    for (WorkflowExecution execution : executions) {
+      for (AbstractMetisPlugin plugin : execution.getMetisPlugins()) {
+        if (!ignoredPluginIds.contains(plugin.getId())) {
+          final ExecutionPluginNode node = new ExecutionPluginNode(execution, plugin);
+          nodesById.put(node.getId(), node);
+        }
+      }
+    }
+
+    // Add all nodes to the lookup map for predecessors
+    final Function<ExecutionPluginNode, PluginLinkByTypeAndTimestamp> createPredecessorKey = node -> new PluginLinkByTypeAndTimestamp(
```

Comment:
That is not a predecessor, this is a nodeByTypeAndTimestamp if I can describe it so..

Reply:
Agreed. Fixed.